### PR TITLE
Test nn.Module on non-contiguous inputs

### PR DESCRIPTION
**Pull request by ssnl**
1. Let `ModuleTest` raise when they fail on non-contiguous inputs. Fix legacy modules.
2. Fix BN (both THNN and cuDNN) not working on non-contiguous inputs.
3. Fix CUDA EmbeddingBag not working on non-contiguous inputs. To prevent calling `.contiguous()` on in both `forward` and `backward`, 
  a. prefix all current `embedding_bag*` functions with `_`, indicating that they require input to be contiguous (there is a check in each function).
  b. create `embedding_bag`, which makes input arguments `.contiguous()`, and calls `_embedding_bag`
3. Make many ATen `embedding*` functions to work on non-contiguous inputs so we don't need to call `input = input.contiguous()` in Python `nn.functional.embedding`.
4. Fix dense-sparse addition when the sparse input is not coalesced and indices or values tensor is not contiguous. This came up in the test cases of Embedding modules with `sparse=True`. Added tests.
5. Update `TensorUtils.cpp` to use `AT_*` macros.

Request:
review from @cpuhrsch on the `Embedding*` changes.
review from @ezyang on ATen sparse & BN changes.

### CODE REVIEW

File: aten/src/ATen/TensorUtils.cpp [hunk 1]
```diff
@@ -20,32 +20,24 @@ std::ostream& operator<<(std::ostream & out, TensorGeometryArg t) {
 }
 
 void checkDim(CheckedFrom c, const TensorGeometryArg& t, int64_t dim) {
-  if (t->dim() != dim) {
```
Comment (ezyang) [inline aten/src/ATen/TensorUtils.cpp]: Very nice

File: aten/src/ATen/native/Embedding.cpp [hunk 2]
```diff
@@ -154,13 +152,13 @@ Tensor & embedding_renorm_cpu_(
     Tensor & self, const Tensor & indices, double max_norm, double norm_type) {
   auto self_arg = TensorArg(self, "self", 1);
   auto indices_arg = TensorArg(indices, "indices", 2);
-  checkContiguous("embedding_renorm_", self_arg);
```
Comment (ezyang) [inline aten/src/ATen/native/Embedding.cpp]: You dropped this test but I don't see a subsequent contiguous() call. Does slicing and then norming work, no matter the stride? I guess there is no reason it shouldn't.

File: aten/src/ATen/native/EmbeddingBag.cpp [hunk 3]
```diff
@@ -67,7 +72,7 @@ static void make_bag_size(const Tensor &offsets, const Tensor &indices,
 static Tensor apply_bag_size(const Tensor &offsets, const Tensor &indices,
                              const int64_t mode, Tensor &output,
                              const Tensor &bag_size) {
-  if (mode == 1) { // MODE_MEAN
+  if (mode == MODE_MEAN) { // MODE_MEAN
```
Comment (ezyang) [inline aten/src/ATen/native/EmbeddingBag.cpp]: nit: no need for the comment anymore :)

File: aten/src/ATen/native/Embedding.cpp (hunk 2 above)
Comment (SsnL) [reply aten/src/ATen/native/Embedding.cpp]: Yeah, since it's a slicing on the tensor, it should just work.

File: test/test_sparse.py [hunk 4]
```diff
@@ -517,6 +520,27 @@ def _test_spadd_shape(self, shape_i, shape_v=None):
 
         self.assertEqual(res, expected)
 
+        x, i, v = self._gen_sparse(len(shape_i), 10, shape)
+        nnz = i.size(1)
+
+        # Non contiguous sparse indices tensor
```
Comment (ezyang) [inline test/test_sparse.py]: I'm not convinced that we actually care about supporting this case, but we can make a further determination on this as we discover how much work it is to keep this working.

**Review by ezyang — APPROVED** [review]

**Review by soumith — APPROVED** [review]

**Comment by SsnL** [comment]
successful rerun: https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-trusty-py2.7.9-test/10128/console

**Pull request closed without merging**